src/program.hpp

```diff
@@ -539,20 +539,124 @@ struct cvk_program : public _cl_program, api_object<object_magic::program> {
         }
     }
 
+private:
+    bool read_llvm_bitcode(const unsigned char* src, size_t size) {
+        if (size >= 4 && (src[0] == 'B' && src[1] == 'C' && src[2] == 0xc0 &&
+                          src[3] == 0xde)) {
+            m_ir.resize(size);
+            memcpy(m_ir.data(), src, size);
+            m_dev_status[m_context->device()] = CL_BUILD_SUCCESS;
+            return true;
+        }
+        return false;
+    }
+
+    const uint32_t clvk_binary_magic =
+        0x6B766C63; // "clvk" in ASCII in little-endian
+    const uint32_t clvk_binary_version = 1;
+    struct clvk_binary_header {
+        uint32_t magic;
+        uint32_t version;
+        uint32_t binary_type;
+    };
+#define COPY_WORD(dst, src)                                                    \
+    do {                                                                       \
+        ((unsigned char*)dst)[0] = ((unsigned char*)(src))[0];                 \
+        ((unsigned char*)dst)[1] = ((unsigned char*)(src))[1];                 \
+        ((unsigned char*)dst)[2] = ((unsigned char*)(src))[2];                 \
+        ((unsigned char*)dst)[3] = ((unsigned char*)(src))[3];                 \
+    } while (0)
+
+    void write_binary_header(unsigned char* dst) const {
+        struct clvk_binary_header* header = (struct clvk_binary_header*)dst;
+        COPY_WORD(&header->magic, &clvk_binary_magic);
+        COPY_WORD(&header->version, &clvk_binary_version);
+        COPY_WORD(&header->binary_type, &m_binary_type);
+    }
+    CHECK_RETURN cl_program_binary_type
+    read_binary_header(const unsigned char* src, size_t size) {
+        struct clvk_binary_header* header = (struct clvk_binary_header*)src;
+        if (size < sizeof(header)) {
+            return CL_PROGRAM_BINARY_TYPE_NONE;
+        }
+        uint32_t magic, version, binary_type;
+        COPY_WORD(&magic, &header->magic);
+        COPY_WORD(&version, &header->version);
+        COPY_WORD(&binary_type, &header->binary_type);
+        if (magic != clvk_binary_magic) {
+            cvk_info_fn("magic not found");
+            return CL_PROGRAM_BINARY_TYPE_NONE;
+        }
+        if (version != clvk_binary_version) {
+            cvk_warn_fn("wrong version");
+            return CL_PROGRAM_BINARY_TYPE_NONE;
+        }
+        return binary_type;
+    }
+
+public:
     CHECK_RETURN bool read(const unsigned char* src, size_t size) {
```

Comment:
Maybe this function and a couple of others are a bit big for the header file. I know there are existing big-ish functions in headers but I'd like to change this where there are no performance or other considerations.